File: api/v1/endpoints/subscriptions.py
# -*- coding: utf-8 -*-
"""Stock subscription endpoints (profile, CRUD, push logs)."""

from __future__ import annotations

import logging

from fastapi import APIRouter, Depends, HTTPException

from api.deps import get_current_user
from api.v1.schemas.subscriptions import (
    NotificationProfileResponse,
    NotificationProfileUpdateRequest,
    SubscriptionCreateRequest,
    SubscriptionItemResponse,
    SubscriptionListResponse,
    SubscriptionPricingResponse,
    SubscriptionPushLogListResponse,
    SubscriptionUpdateRequest,
)
from src.services.subscription_service import SubscriptionService, SubscriptionValidationError
from src.user_context import CurrentUser

logger = logging.getLogger(__name__)

router = APIRouter()


def _service() -> SubscriptionService:
    return SubscriptionService.get_instance()


def _validation_error(exc: SubscriptionValidationError) -> HTTPException:
    return HTTPException(
        status_code=400,
        detail={"error": "validation_error", "message": str(exc)},
    )


@router.get(
    "/profile",
    response_model=NotificationProfileResponse,
    summary="Get notification profile",
)
async def get_notification_profile(current_user: CurrentUser = Depends(get_current_user)):
    payload = _service().get_profile(current_user.id)
    return NotificationProfileResponse.model_validate(payload)


@router.put(
    "/profile",
    response_model=NotificationProfileResponse,
    summary="Update notification profile",
)
async def update_notification_profile(
    body: NotificationProfileUpdateRequest,
    current_user: CurrentUser = Depends(get_current_user),
):
    try:
        payload = _service().save_profile(
            current_user.id,
            notification_email=body.notification_email,
            webhook_urls=body.webhook_urls,
            webhook_bearer_token=body.webhook_bearer_token,
            clear_bearer_token=body.clear_webhook_bearer_token,
        )
    except SubscriptionValidationError as exc:
        raise _validation_error(exc) from exc
    return NotificationProfileResponse.model_validate(payload)


@router.get(
    "/pricing",
    response_model=SubscriptionPricingResponse,
    summary="Get subscription pricing",
)
async def get_subscription_pricing(current_user: CurrentUser = Depends(get_current_user)):
    _ = current_user
    return SubscriptionPricingResponse.model_validate(_service().get_pricing())


@router.get(
    "",
    response_model=SubscriptionListResponse,
    summary="List stock subscriptions",
)
async def list_subscriptions(current_user: CurrentUser = Depends(get_current_user)):
    payload = _service().list_subscriptions(current_user.id)
    return SubscriptionListResponse.model_validate(payload)


@router.post(
    "",
    response_model=SubscriptionItemResponse,
    summary="Create stock subscription",
)
async def create_subscription(
    body: SubscriptionCreateRequest,
    current_user: CurrentUser = Depends(get_current_user),
):
    try:
        payload = _service().create_subscription(
            current_user.id,
            code=body.code,
            name=body.name,
            interval_days=body.interval_days,
        )
    except SubscriptionValidationError as exc:
        raise _validation_error(exc) from exc
    return SubscriptionItemResponse.model_validate(payload)


@router.get(
    "/push-logs",
    response_model=SubscriptionPushLogListResponse,
    summary="List recent subscription push logs",
)
async def list_subscription_push_logs(
    limit: int = 20,
    current_user: CurrentUser = Depends(get_current_user),
):
    items = _service().list_push_logs(current_user.id, limit=max(1, min(limit, 100)))
    return SubscriptionPushLogListResponse.model_validate({"items": items})


@router.patch(
    "/{subscription_id}",
    response_model=SubscriptionItemResponse,
    summary="Update stock subscription",
)
async def update_subscription(
    subscription_id: int,
    body: SubscriptionUpdateRequest,
    current_user: CurrentUser = Depends(get_current_user),
):
    try:
        payload = _service().update_subscription(
            current_user.id,
            subscription_id,
            interval_days=body.interval_days,
            status=body.status,
        )
    except SubscriptionValidationError as exc:
        raise _validation_error(exc) from exc
    return SubscriptionItemResponse.model_validate(payload)


@router.delete(
    "/{subscription_id}",
    summary="Delete stock subscription",
)
async def delete_subscription(
    subscription_id: int,
    current_user: CurrentUser = Depends(get_current_user),
):
    try:
        _service().delete_subscription(current_user.id, subscription_id)
    except SubscriptionValidationError as exc:
        raise _validation_error(exc) from exc
    return {"success": True}

File: apps/dsa-web/src/api/subscriptions.ts
import apiClient from './index';
import { toCamelCase } from './utils';
import type {
  NotificationProfile,
  SubscriptionItem,
  SubscriptionListResponse,
  SubscriptionPricing,
  SubscriptionPushLogListResponse,
} from '../types/subscriptions';

export const subscriptionsApi = {
  async getProfile(): Promise<NotificationProfile> {
    const response = await apiClient.get<Record<string, unknown>>('/api/v1/subscriptions/profile');
    return toCamelCase(response.data) as NotificationProfile;
  },

  async saveProfile(payload: {
    notificationEmail: string;
    webhookUrls: string;
    webhookBearerToken?: string;
    clearWebhookBearerToken?: boolean;
  }): Promise<NotificationProfile> {
    const response = await apiClient.put<Record<string, unknown>>('/api/v1/subscriptions/profile', {
      notification_email: payload.notificationEmail,
      webhook_urls: payload.webhookUrls,
      webhook_bearer_token: payload.webhookBearerToken,
      clear_webhook_bearer_token: payload.clearWebhookBearerToken ?? false,
    });
    return toCamelCase(response.data) as NotificationProfile;
  },

  async getPricing(): Promise<SubscriptionPricing> {
    const response = await apiClient.get<Record<string, unknown>>('/api/v1/subscriptions/pricing');
    return toCamelCase(response.data) as SubscriptionPricing;
  },

  async list(): Promise<SubscriptionListResponse> {
    const response = await apiClient.get<Record<string, unknown>>('/api/v1/subscriptions');
    return toCamelCase(response.data) as SubscriptionListResponse;
  },

  async create(payload: {
    code: string;
    name?: string;
    intervalDays: number;
  }): Promise<SubscriptionItem> {
    const response = await apiClient.post<Record<string, unknown>>('/api/v1/subscriptions', {
      code: payload.code,
      name: payload.name,
      interval_days: payload.intervalDays,
    });
    return toCamelCase(response.data) as SubscriptionItem;
  },

  async update(
    subscriptionId: number,
    payload: { intervalDays?: number; status?: 'active' | 'paused' },
  ): Promise<SubscriptionItem> {
    const response = await apiClient.patch<Record<string, unknown>>(
      `/api/v1/subscriptions/${subscriptionId}`,
      {
        interval_days: payload.intervalDays,
        status: payload.status,
      },
    );
    return toCamelCase(response.data) as SubscriptionItem;
  },

  async remove(subscriptionId: number): Promise<void> {
    await apiClient.delete(`/api/v1/subscriptions/${subscriptionId}`);
  },

  async listPushLogs(limit = 20): Promise<SubscriptionPushLogListResponse> {
    const response = await apiClient.get<Record<string, unknown>>(
      '/api/v1/subscriptions/push-logs',
      { params: { limit } },
    );
    return toCamelCase(response.data) as SubscriptionPushLogListResponse;
  },
};
